src/ucs/memory/numa.c

```diff
@@ -72,3 +88,228 @@ int ucs_numa_node_of_cpu(int cpu)
+    }
+
+    if ((parsed_node < 0) || (parsed_node > UCS_NUMA_MAX_NODE)) {
+        ucs_error("Failed to parse sysfs node fd");
+        return UCS_NUMA_DEFAULT_NODE;
+    }
+
+    return parsed_node;
+}
+
+/**
+ * Parse and iterate the distance list of @param source.
+ * For each neighbor of source add distance(source, neighbor)
+ * to numa_distance_hash.
+ *
+ * @param [in]  distances Numa distances list string
+ * @param [in]  source source numa node
+ * @param [in]  dest destination numa node
+ *
+ * @return Distance from @param source to @param dest
+ */
+static ucs_numa_distance_t
+ucs_numa_node_insert_distances(char *distances, ucs_numa_node_t source,
+                               ucs_numa_node_t dest)
+{
+    char *token = strtok(distances, UCS_NUMA_NODE_DISTANCE_DELIMITER);
+    ucs_numa_distance_t distance_to_dest = UCS_NUMA_DEFAULT_DISTANCE;
+    ucs_numa_node_t node                 = 0;
+    ucs_numa_distance_t distance;
+    ucs_kh_put_t kh_put_status;
+    khiter_t hash_it;
+
+    while (token) {
+        distance = strtoul(token, 0, 0);
+        ucs_assert(distance >= UCS_NUMA_MIN_DISTANCE);
+        if (node == dest) {
+            distance_to_dest = distance;
+        }
+
+        hash_it = kh_put(numa_distance /*name*/,
+                         &numa_distance_hash /*pointer to hashmap*/,
+                         ucs_numa_distance_hash_key(source, node) /*key*/,
+                         &kh_put_status);
+        if ((kh_put_status != UCS_KH_PUT_FAILED) &&
+            (kh_put_status != UCS_KH_PUT_KEY_PRESENT)) {
+            kh_value(&numa_distance_hash, hash_it) = distance;
+        }
+
+        node++;
+        token = strtok(NULL, UCS_NUMA_NODE_DISTANCE_DELIMITER);
+    }
+
+    return distance_to_dest;
+}
+
+ucs_numa_distance_t
+ucs_numa_distance(ucs_numa_node_t source, ucs_numa_node_t dest)
+{
+    ucs_numa_distance_t distance = UCS_NUMA_DEFAULT_DISTANCE;
+    size_t buf_len               = 0;
+    ssize_t len                  = 0;
+    char *line                   = NULL;
+    char node_distance_fd_path[PATH_MAX];
+    FILE *distance_fd;
+    khiter_t hash_it;
+
+    ucs_assert(source <= ucs_numa_get_max_node());
+    ucs_assert(dest <= ucs_numa_get_max_node());
+
+    hash_it = kh_get(numa_distance /*name*/,
+                     &numa_distance_hash /*pointer to hashmap*/,
+                     ucs_numa_distance_hash_key(source, dest) /*key*/);
+
+    if (hash_it != kh_end(&numa_distance_hash)) {
+        distance = kh_value(&numa_distance_hash, hash_it);
+    } else {
+        ucs_snprintf_safe(node_distance_fd_path, PATH_MAX,
+                          UCS_NUMA_NODE_DISTANCE_FD_PATH, source);
+        distance_fd = fopen(node_distance_fd_path, "r");
+
+        if (distance_fd) {
+            len = getdelim(&line, &buf_len, '\n', distance_fd);
+            fclose(distance_fd);
+            if (len > 0) {
+                distance = ucs_numa_node_insert_distances(line, source, dest);
+                if (line) {
```

Comment:
Please check whether there is a case when `len == 0` but `line != NULL`, leading to a memory leak here.